src/hooks/useTimer.ts

```diff
@@ -1,33 +1,39 @@
-import { useState, useEffect } from 'react';
+import { useState, useEffect, useRef, useCallback } from 'react';
 
 export function useTimer(initialTime: number, callback: () => void) {
-  const [timeLeft, setTimeLeft] = useState<number>(0);
-  const [timerId, setTimerId] = useState<NodeJS.Timeout | null>(null);
+  const [timeLeft, setTimeLeft] = useState<number>(initialTime);
+  const [isRunning, setIsRunning] = useState<boolean>(false);
+  const timerRef = useRef<NodeJS.Timeout | null>(null);
+  const callbackRef = useRef(callback);
 
-  const startTimer = () => {
-    setTimeLeft(initialTime);
-    if (timerId) clearInterval(timerId);
-
-    const newTimerId = setInterval(() => {
-      setTimeLeft((prevTime) => {
-        if (prevTime <= 1) {
-          clearInterval(newTimerId);
-          setTimerId(null);
-          callback();
-          return 0;
-        }
-        return prevTime - 1;
-      });
-    }, 1000);
+  useEffect(() => {
+    callbackRef.current = callback;
+  }, [callback]);
 
-    setTimerId(newTimerId);
-  };
+  const startTimer = useCallback(() => {
+    setIsRunning(true);
+    setTimeLeft(initialTime);
+  }, [initialTime]);
 
   useEffect(() => {
+    if (isRunning && timeLeft > 0) {
+      timerRef.current = setTimeout(() => {
+        setTimeLeft((prevTime) => prevTime - 1);
+      }, 1000);
+    } else if (isRunning && timeLeft === 0) {
+      setIsRunning(false);
+      if (callbackRef.current) callbackRef.current();
```

Comment:
여기 조건문 없어도 되지 않을까요??
```suggestion
      callbackRef.current();
```

Reply:
이 훅에서는 컴포넌트가 마운트될 때 callback이 제공되어 굳이 필요 없긴 하네요!
습관 상 넣었던 것 같습니다

✅ 5e9374f27dbd33b642d85dabf66cc8ef2936ec99